### DISCUSSION TITLE
Does pyo3 automatically release GIL if py is not used or do you have to do this manually?

### DISCUSSION THREAD

**Discussion by rafales** [PyO3/pyo3#3621] [Questions] [answered]
Quick question. I've been looking at code of a python package. Code is fairly simple and it's here:

https://github.com/mgd020/mjml-python/blob/master/src/lib.rs

One thing I don't understand about PyO3: given that mjml2html doesn't hold any references to `Python` - will this code run without GIL or do you need to call `allow_threads` explicitly in `mjml2html`?

**Comment by adamreichold** [comment, marked as answer]
The linked code with hold the GIL for the whole duration of the call. If you want to release it, you to do so manually by calling `allow_threads`.

(We are working an a function attribute `#[pyo3(allow_threads)]` in #3610 to do this automatically, but that code is neither merged nor released yet.)
